### DISCUSSION TITLE
Long execution times

### DISCUSSION THREAD

**Discussion by luisartg** [pnp/pnpframework#341] [Q&A]
My current template takes more than 120 minutes to be applied to a new site.
It contains 15 Document LIbraries, each one with around 4 views and 53 field references.

Is it normal that it takes that amount of time?
We are currently using wsp templates and it takes around 10 min to create the whole site. However, these templates are no longer supported for Modern sites.

Aside from the problem that a process that took around 10 min now takes 2 hours, we are currently migrating all our processes to the cloud, and I'm not sure an azure function can last 2 hours executing: could be possible to divide the template in various sub-templates, each part taking care of a minimal task. For example: one template creates a library, another creates the views, other the fields references and another one the security. That way the whole process may take the 2 hours, but is divided in activities of around 5 minutes.

**Comment by jansenbe** [comment]
@Luis-Art-Guerra : that's indeed quite long, not what I would expect. Might be that you're getting throttled? 

You can indeed split the templates in smaller parts, each doing one feature (e.g fields / lists / ....) or each doing one entity (e.g. doc lib 1 with all dependencies, then doc lib 2, ...)

Would also be good if you could figure out which handler (lists, fields, ...) is so slow, feels that there's a performance improvement. If you find something please open an issue in the issue list for it

**Reply by luisartg** [reply]
Hi @jansenbe, i added the debug output to a gist if you would like to take a look :)
https://gist.github.com/Luis-Art-Guerra/360c945a0298f13066e8fe9a1a0718b8

From what I've seen, there is no reported retries or throttling. At the beginning there is the report of the following missing features:
95e9819c-3d2b-45cb-8faf-5e1dadbcba17
00c0ef53-c1d1-4abd-aff3-1ee66b76abc5
8cef1561-00d2-45bc-a2da-b0b1dddf7400
b1e186bd-450a-4629-ad26-e27fb7a9ca04
79fa51bd-3b9e-4beb-a6ee-f956c30a7677
59cbf073-76c8-4bf6-9966-fdd9684e1f27

I have no idea which features are those, I can't find the references. I'll do another run trying to add those before applying the template.

As for the rest of the execution seems pretty normal. Maybe my template has too many things to add?
Some of the times the process is taking are these:
Start time - End time - total minutes - Type - Quantity
22:49:32.3681 - 22:59:12.6508 - 9:40 - Create Lists - 16
22:59:25.0549 - 23:26:52.2917 - 27:27 - To Update All field references of all libraries - 805 
23:28:26.3966 - 00:01:18.4718 - 32:52 - Updating Lookup fields for all libraries - 825
00:01:23.2228 - 00:34:37.1176 - 33:14 - List Settings - 877 - This activity ended incomplete due to an error

I hope this info can help. For the time being, i'll split the template to minimize execution time for the azure functions and faster debugging.

**Reply by luisartg** [reply]
To add more information, this is happening also with PnPFramework for powershell. Same behavior, similar times.
